### PR TITLE
Feature/index scan limit push down

### LINKED ISSUE #2950: [Feature] index scan limit push down
from issue #2602 

Related PR:
#2905 
#2823 
#2796

### PR DESCRIPTION
**Pull request by Shylock-Hg**
Require https://github.com/vesoft-inc/nebula/pull/2796

Sub job of #2602

### CODE REVIEW

**Comment by bright-starry-sky** [comment]
please resolve the conflicts.

**Comment by Shylock-Hg** [comment]
> please resolve the conflicts.

Done.

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    OptContext *octx, const MatchedResult &matched) const {
+  auto limitGroupNode = matched.node;
+  auto projGroupNode = matched.dependencies.front().node;
+  auto indexScanGroupNode = matched.dependencies.front().dependencies.front().node;
+
+  const auto limit = limitGroupNode->node()->asNode<const Limit>();
+  const auto proj = projGroupNode->node()->asNode<const Project>();
+  const auto indexScan = indexScanGroupNode->node()->asNode<const IndexScan>();
+
+  int64_t limitRows = limit->offset() + limit->count();
```
Comment (bright-starry-sky) [inline src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:53]: Could you give me a comment? What does "offset" mean? 
for example `limit 5 offset 3` , Starting at line 3, continue to get 5 rows. the total limit is 5.

Comment (bright-starry-sky) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:53]: I think `offset` is not easy to handle in IndexScan

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:53]: The sentence `limit 3, 5` means get 5 elements from the third element.

**Comment by bright-starry-sky** [comment]
I didn't find the keyword LIMIT in lookup_sentence, did you solve it via other way?

**Comment by Shylock-Hg** [comment]
> I didn't find the keyword LIMIT in lookup_sentence, did you solve it via other way?

In the compound sentence `lookup .. | limit 3, 5`

**Comment by bright-starry-sky** [comment]
> > I didn't find the keyword LIMIT in lookup_sentence, did you solve it via other way?
> 
> In the compound sentence `lookup .. | limit 3, 5`

I see. our ideas seem to diverge，let me explain what I think.I'm not sure what I think is right. It's just a discussion.
for example :
part1 ---> 10 rows;
part2 ---> 10 rows;

- lookup tag ... limit 5 ---> The number of rows returned should be 5 + 5 == 10 (part1 = 5 rows, part2 = 5 rows)
- lookup tag ... | limit 5 ---> The number of rows returned should be 5 (part1 = 10 rows, part2 = 10 rows, need to truncate it to 5  rows at the graph layer)
- lookup tag ... limit 5 | limit 6 ---> The number of rows returned should be 5 + 5 == 10 (part1 = 5 rows, part2 = 5 rows. need to truncate it to 6  rows at the graph layer)

**Comment by Shylock-Hg** [comment]
> > > I didn't find the keyword LIMIT in lookup_sentence, did you solve it via other way?
> > 
> > 
> > In the compound sentence `lookup .. | limit 3, 5`
> 
> I see. our ideas seem to diverge，let me explain what I think.I'm not sure what I think is right. It's just a discussion.
> for example :
> part1 ---> 10 rows;
> part2 ---> 10 rows;
> 
> * lookup tag ... limit 5 ---> The number of rows returned should be 5 + 5 == 10 (part1 = 5 rows, part2 = 5 rows)
> * lookup tag ... | limit 5 ---> The number of rows returned should be 5 (part1 = 10 rows, part2 = 10 rows, need to truncate it to 5  rows at the graph layer)
> * lookup tag ... limit 5 | limit 6 ---> The number of rows returned should be 5 + 5 == 10 (part1 = 5 rows, part2 = 5 rows. need to truncate it to 6  rows at the graph layer)

Fixed.

**Review by bright-starry-sky — DISMISSED** [review]
Good job. LGTM.

File: src/graph/planner/plan/PlanNode.h [hunk 2]
```diff
@@ -15,169 +15,206 @@
...
+  }
+
+  // Variant X is Variant X
+  // Variant X is main type of Variant X
+  // Variant X is not Variant Y
+  bool isA(Kind kind) const {
+    if (kind == kind_) {
+      return true;
+    }
+    if (variantType(kind) != 0) {
```
Comment (critical27) [inline src/graph/planner/plan/PlanNode.h:209]: Consider add a macro to indicate 0 as a base_type

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp (hunk 1 above)
Comment (critical27) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:53]: @bright-starry-sky do we use `offset` to paging? Cursor will be a better choice.

**Review by critical27 — COMMENTED** [review]
Generally LGTM, good job. Ready to merge for now?

**Comment by critical27** [comment]
@CPWstatic @yixinglu, PTAL.

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp (hunk 1 above)
Comment (bright-starry-sky) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:53]: > @bright-starry-sky do we use `offset` to paging? Cursor will be a better choice.

No, we did not use `offset`, the current approach is similar to a cursor.

Comment (bright-starry-sky) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:53]: > @bright-starry-sky do we use `offset` to paging? Cursor will be a better choice.

please ignore the concept of `offset`. It has nothing to do with index scanning.

File: tests/tck/features/lookup/LookUpLimit.feature [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      | id | name               | dependencies | operator info  |
+      | 3  | DataCollect        | 4            |                |
+      | 4  | Sort               | 5            |                |
+      | 5  | Limit              | 6            |                |
+      | 6  | Project            | 7            |                |
+      | 7  | TagIndexPrefixScan | 8            | {"limit": "3"} |
+      | 8  | Start              |              |                |
+    When profiling query:
+      """
+      LOOKUP ON like WHERE like.likeness == 90 Limit 3 | LIMIT 3 | ORDER BY $-.SrcVID
```
Comment (czpmango) [inline tests/tck/features/lookup/LookUpLimit.feature:161]: Will we support a similar syntax for go clause, just like:
```
go from "Tim Duncan" over like yield like._dst as dst limit 3 | LIMIT 3
LOOKUP ON player Limit 3 | LIMIT 3
```

Comment (Shylock-Hg) [reply tests/tck/features/lookup/LookUpLimit.feature:161]: Go will support the step sample/limit

**Review by critical27 — DISMISSED** [review]

**Comment by yixinglu** [comment]
> @CPWstatic @yixinglu, PTAL.

@critical27  @bright-starry-sky 

I don't agree with the changes about the Plan Node kind, this new definition is quite ugly. I prefer to meet the needs by enhancing the flexibility of pattern, rather than by modifying the way plan node's kind is defined.

**Comment by Shylock-Hg** [comment]
> > @CPWstatic @yixinglu, PTAL.
> 
> @critical27 @bright-starry-sky
> 
> I don't agree with the changes about the Plan Node kind, this new definition is quite ugly. I prefer to meet the needs by enhancing the flexibility of pattern, rather than by modifying the way plan node's kind is defined.

How to enhance the pattern?

**Comment by yixinglu** [comment]
> > > @CPWstatic @yixinglu, PTAL.
> > 
> > 
> > @critical27 @bright-starry-sky
> > I don't agree with the changes about the Plan Node kind, this new definition is quite ugly. I prefer to meet the needs by enhancing the flexibility of pattern, rather than by modifying the way plan node's kind is defined.
> 
> How to enhance the pattern?

Try one rule to match multiple patterns

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+std::unique_ptr<OptRule> PushLimitDownIndexScanRule::kInstance =
+    std::unique_ptr<PushLimitDownIndexScanRule>(new PushLimitDownIndexScanRule());
+
+PushLimitDownIndexScanRule::PushLimitDownIndexScanRule() { RuleSet::QueryRules().addRule(this); }
+
+const Pattern &PushLimitDownIndexScanRule::pattern() const {
+  static Pattern pattern = Pattern::create(
+      graph::PlanNode::Kind::kLimit,
+      {Pattern::create(graph::PlanNode::Kind::kProject,
+                       {Pattern::create(graph::PlanNode::Kind::kIndexScan, {}, true)})});
```
Comment (yixinglu) [inline src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:39]: here should not be one opt rule, but should be split into two rules: limit <-> project, limit -> indexscan

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:39]: Why need limit->indexscan?

**Comment by Shylock-Hg** [comment]
> > > > @CPWstatic @yixinglu, PTAL.
> > > 
> > > 
> > > @critical27 @bright-starry-sky
> > > I don't agree with the changes about the Plan Node kind, this new definition is quite ugly. I prefer to meet the needs by enhancing the flexibility of pattern, rather than by modifying the way plan node's kind is defined.
> > 
> > 
> > How to enhance the pattern?
> 
> Try one rule to match multiple patterns

I think it's ok. We refactor when get better way in future.

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp (hunk 4 above)
Comment (yixinglu) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:39]: Push limit down indexscan

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:39]: The sentence only generate `limit`->`project`->`indexscan` plan

**Comment by bright-starry-sky** [comment]
What's new advances ? @Shylock-Hg @yixinglu .
If some differences require further discussion. would be better to split the current PR into two?

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp (hunk 4 above)
Comment (yixinglu) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:39]: You should consider the optimization rules from the perspective of the optimizer, not just for this case.

**Comment by yixinglu** [comment]
> What's new advances ? @Shylock-Hg @yixinglu .
> If some differences require further discussion. would be better to split the current PR into two?

I don’t accept refactoring in the future. If don’t want to optimize the pattern, then implement a few more similar optimization rules.

**Comment by Shylock-Hg** [comment]
> > What's new advances ? @Shylock-Hg @yixinglu .
> > If some differences require further discussion. would be better to split the current PR into two?
> 
> I don’t accept refactoring in the future. If don’t want to optimize the pattern this time, then implement a few more similar optimization rules.

In fact, I'm not sure about refactoring. Match variants at once is reasonable demand in this optimizer. Write the multiple almost similar rules is not well.

File: src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp (hunk 4 above)
Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp:39]: We don't need to match the nonexistent pattern.

**Comment by Shylock-Hg** [comment]
Reopen in https://github.com/vesoft-inc/nebula/pull/2966

**Pull request closed without merging**